File: ruby/binary-search-tree/binary_search_tree.rb
class Bst

  attr_reader :data
  attr_accessor :left, :right
  def initialize(data)
    @data = data
    @left = nil
    @right = nil
  end

  def insert(new_datum, bst = self)
    if new_datum <= bst.data && !bst.left
      bst.left = Bst.new(new_datum)
    elsif new_datum > bst.data && !bst.right
      bst.right = Bst.new(new_datum)
    elsif new_datum <= bst.data
      insert(new_datum, bst.left)
    else
      insert(new_datum, bst.right)
    end
  end

  def each(&block)
    flattened.each(&block)
  end

  def flattened(bst = self)
    if !bst
      []
    else
      flattened(bst.left) + [bst.data] + flattened(bst.right)
    end
  end
end

File: javascript/binary-search-tree/binary-search-tree.js
'use strict'

class BinarySearchTree {
  constructor(data) {
    this.data = data;
    this.right = null;
    this.left = null;
  }

  data () {
    this.data;
  }

  insert(data) {
    if (!this.data) {this.data = data;}
    if (data > this.data) {
      if (this.right) {
        this.right.insert(data);
      } else {
        this.right = new BinarySearchTree(data)
      }
    } else if (data <= this.data) {
      if (this.left) {
        this.left.insert(data);
      } else {
        this.left = new BinarySearchTree(data)
      }
    }
  }

  each(foo) {
    if (!this) {
      return;
    } else {
      if (this.left) { this.left.each(foo); }
      foo(this.data);
      if (this.right) { this.right.each(foo);}
    }
    return;
  }
}



module.exports = BinarySearchTree;
